### ISSUE TITLE
About gradient accumulation

### ISSUE THREAD

**Issue by bobo0810** [davda54/sam#3]
@davda54 Thank you for providing the implementation of pytorch. The following is about the implementation of gradient accumulation. I am not sure whether the code is correct or not, so I hope to get your help~ 
![93322be129d1a3aecc4a178eba549c4](https://user-images.githubusercontent.com/26057879/97521415-7dd9fb00-19d8-11eb-8e93-2d8fe025c4b4.png)

**Comment by davda54** [comment]
Hi, I believe that's the right way of doing gradient accumulation. Indeed, it is a bit cumbersome, but I can't think of a more elegant way to do it. Is it working for you?

**Comment by bobo0810** [comment]
> Hi, I believe that's the right way of doing gradient accumulation. Indeed, it is a bit cumbersome, but I can't think of a more elegant way to do it. Is it working for you?

Thank you for your reply. I'm sorry I haven't started the experiment yet. If I finish the experiment, I will post it as soon as possible.

**Comment by bobo0810** [comment]
@davda54 Hi, I did a toy experiment with 10,000 categories. Under the same parameters, SAM performs better than SGD!

**Comment by davda54** [comment]
Awasome! :) To be honest, I have no experience with AMP, but I don't see any major issues in its combination with SAM.

**Comment by ODD2** [comment]
Hi @davda54 , I saw a recent [discussion](https://github.com/davda54/sam/issues/62) about gradient accumulation with the SAM optimizer and got a bit confused, so would like to make sure if the current SAM is still adaptable with the gradient accumulation method?
